### PR TITLE
DS-459 Add Dialog component

### PR DESCRIPTION
**Pull request by danielamorse**
## Jira

https://pegadigitalit.atlassian.net/browse/DS-459

## Summary

Add new experimental Dialog component. It is _not_ officially ready to be used.

## Details

Adds new Dialog component. It is not a web component. It includes support for the same props as Bolt Modal except for scroll behavior (dialog container scrolls by default).

This uses the new `<dialog>` element and its associated browser API. It is supported by all modern browsers, but for Safari it is only supported in 15.4 and up. We are merging this work now but marking in the docs that it is not officially released until adoption of the latest versions of Safari improves.

There is no JS documentation in PL yet because you can not easily call the Dialog methods from outside of Bolt. This is because the JS in Bolt is loaded asynchronously, which means you need to also add some event listener to know if/when the component has been loaded, and it quickly gets complicated. I'm going to look into an alternate approach using [Drupal libraries and behaviors](https://pegadigitalit.atlassian.net/browse/DS-783).

One other known issue is the close button positioning, see https://pegadigitalit.atlassian.net/browse/DS-782.

## How to test

- Review code and docs

### Dialog
- Dialog component is added. It is _not_ ready for sites to use until Safari 15.4 adoption improves.

### CODE REVIEW

**Comment by colbytcook** [comment]
@danielamorse I am not seeing the `content` description being displayed for the "Dialog Header", "Dialog Body", or "Dialog Footer" schema section in Patterrnlab. Not sure if we care about this or not but I figured I would mention it.

File: docs-site/src/pages/pattern-lab/_patterns/40-components/dialog/-05-dialog.twig [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  },
+} only %}
+{% endverbatim %}{% endset %}
+
+{% include '@utils/pattern-doc-page.twig' with {
+  title: 'Basic Dialog',
+  description: description,
+  notes: notes,
+  demo: demo,
+  twig_markup: twig_markup,
```
Comment (colbytcook) [inline docs-site/src/pages/pattern-lab/_patterns/40-components/dialog/-05-dialog.twig:78]: @danielamorse I am not seeing this markup example show in Patternlab

**Comment by colbytcook** [comment]
@danielamorse I am seeing a mobile video display error, see the attached screenshot.
mobile-video-error

**Comment by danielamorse** [comment]
> @danielamorse I am seeing a mobile video display error, see the attached screenshot. mobile-video-error

@mikemai2awesome the dialog has no inline size until `small` breakpoint. Should it be set to 100% [here](https://github.com/boltdesignsystem/bolt/pull/2492/files#diff-8544735a8ef8b261f6071d7b6daa7d2e789cf4ea28199febca2d5b56c17a405eR105-R111)?

**Comment by danielamorse** [comment]
> @danielamorse the close button on an image modal is hard to see, see screenshot. image-close-bug

@colbytcook this will be addressed in https://pegadigitalit.atlassian.net/browse/DS-782.

File: docs-site/src/pages/pattern-lab/_patterns/40-components/dialog/-20-dialog-persistent.twig [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+{% endset %}
+
+{% include '@bolt-components-dialog/dialog.twig' with {
+  content: content,
+  persistent: true,
+  attributes: {
+    id: 'dialog-persistent-1'
+  },
+} only %}
+{% endset %}
```
Comment (colbytcook) [inline docs-site/src/pages/pattern-lab/_patterns/40-components/dialog/-20-dialog-persistent.twig:64]: Is this an extra {% endset %}?

Comment (colbytcook) [reply docs-site/src/pages/pattern-lab/_patterns/40-components/dialog/-20-dialog-persistent.twig:64]: disregard, this is the closing demo set tag, the indents messed me up

**Comment by colbytcook** [comment]
I am seeing a bug on the "/pattern-lab/?p=tests-dialog-in-carousel" page. If you click to the second slide and open tthe carousel the modal will be shown over the first slide and not the second (if  you click back the first slide will see a cropped version of the modal)

File: packages/components/bolt-dialog/src/dialog-footer.twig [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+{# Variables #}
+{% set this = init(schema) %}
+{% set attributes = create_attribute(attributes|default({})) %}
+
+{# Array of classes based on the defined + default props #}
+{% set classes = [
+  'c-bolt-dialog__footer',
+] %}
+
+<div {{ attributes.addClass(classes) }}>
```
Comment (colbytcook) [inline packages/components/bolt-dialog/src/dialog-footer.twig:15]: @danielamorse Should this be wrapped in a <footer> like the header is wrapped in a <header>?

**Comment by danielamorse** [comment]
@colbytcook
**Mobile video display**
Nice catch. Fixed, revealed another issue we can handle separately (https://pegadigitalit.atlassian.net/browse/DS-784).

**Carousel Modal**
The second slide is the old Modal. It is expected to break. I updated the button text so that's more clear.

**Footer**
I'm not sure. The docs lead me to think "it's not that kind of footer". Then again, we _are_ using `<header>` here. @mikemai2awesome?

**Review by colbytcook — APPROVED** [review]
@danielamorse with the two follow tickets (video on mobile and image close button visibility) every looks good to me. Approved

**Review by mikemai2awesome — APPROVED** [review]
Ok to merge. Follow up tickets will take care of remaining issues.

**Pull request merged**